### PR TITLE
[BEAM-7534] add --mountTempDir option for Docker container

### PR DESCRIPTION
**Pull request by ihji**
add --mountTempDir option for easier data sharing with Docker container. This will enable us to test FnApi features without GCS connectivity.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
run java precommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java [hunk 1]
```diff
@@ -32,6 +32,14 @@
 
   void setRetainDockerContainers(boolean retainDockerContainers);
 
+  @Description(
+      "Mount temporary directory to Docker container instance. "
+          + "The parameter is a path relative to the system temporary directory "
+          + "e.g. '--mountTempDir foo' implies '/tmp/foo' on Linux.")
+  String getMountTempDir();
```
Comment (angoenka) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java]: Lets also specify where the mount will be available in docker. 
We can standardize the name path in docker to `/tmp/beam_mount/<mount_dir>`
example `/tmp/beam_mount/tmp/foo`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflector.java [hunk 2]
```diff
@@ -36,7 +36,8 @@ private PipelineOptionsReflector() {}
    *
    * @see PipelineOptionsReflector#getOptionSpecs(Iterable)
    */
-  static Set<PipelineOptionSpec> getOptionSpecs(Class<? extends PipelineOptions> optionsInterface) {
```
Comment (angoenka) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflector.java]: Is this change needed for supporting mountTempDir?

Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflector.java]: Yes. Otherwise mountTempDir option will not be exposed to Python SDK since it's marked as hidden.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java (hunk 1 above)
Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java]: Wouldn't it be better to keep paths as close as possible like /tmp/foo (local) and /tmp/foo (docker)? It might be confusing if we mount /tmp/foo (local) to /tmp/beam_mount/tmp/foo (docker).

**Comment by ihji** [comment]
Run java precommit

**Comment by ihji** [comment]
Retest this please

**Comment by ihji** [comment]
CC: @mxm

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java]: What about conflicts (e.g. if the directory already exists in the docker image). Also, should we allow specifying multiple mount dirs?

Comment (ihji) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/ManualDockerEnvironmentOptions.java]: We can put a unique prefix for helping avoid possible conflict but I think still a rule of thumb is that users need to be cautious in choosing mounting directory name. In case of this new option, the base directory is fixed to /tmp so conflicts might not easily happen.

For multiple mount dirs, I'm not sure how much it's useful since users can create as many directories as they want in a single mounting directory.

**Comment by chadrik** [comment]
@ihji there’s another PR that covers this same topic over at https://nuget.pkg.github.com/apache/beam/pull/8982. It seems like it covers more cases and does so without adding a docker-specific flag.  Do you think it will work sufficiently for your use case?

**Comment by ihji** [comment]
@chadrik I think that PR would cover this use-case. Thanks for the link.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflector.java [hunk 3]
```diff
@@ -36,7 +36,8 @@ private PipelineOptionsReflector() {}
...
+  static Set<PipelineOptionSpec> getOptionSpecs(
+      Class<? extends PipelineOptions> optionsInterface, boolean skipHidden) {
```
Comment (mxm) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflector.java]: This introduces a new feature: hidden options. I'm not sure about the consequences of such a feature. I believe such options would be better stored in the docker environment config.
